Add whitespace handling methods to MemoryExtensions

AlexRadch opened the pull request:
Part of #77959 except `ContainsAnyExceptWhiteSpace`

Updated the MemoryExtensions class with new methods for efficient whitespace handling, including ContainsAnyWhiteSpace, IndexOfAnyWhiteSpace, and LastIndexOfAnyWhiteSpace.

Optimized the IsWhiteSpace method to utilize SearchValues.WhiteSpaces.

Added corresponding test cases in StringSearchValues.cs to validate the new functionality.

AlexRadch committed:
Add whitespace handling methods to MemoryExtensions

Updated the MemoryExtensions class with new methods for
efficient whitespace handling, including ContainsAnyWhiteSpace,
IndexOfAnyWhiteSpace, and LastIndexOfAnyWhiteSpace.

Optimized the IsWhiteSpace method to utilize SearchValues.WhiteSpaces.

Added corresponding test cases in StringSearchValues.cs to validate
the new functionality.

AlexRadch commented:
@stephentoub I have implemented the `MemoryExtensions.LastIndexOfAnyWhiteSpace` method and added a corresponding test, `StringSearchValuesTests.LastIndexOfAnyWhiteSpace_Found`, which currently fails when validating this method.

The `LastIndexOfAnyWhiteSpace` method is entirely based on `SearchValues.LastIndexOfAny`. It's possible that my test is uncovering a bug in the `SearchValues.LastIndexOfAny` method, as I haven't been able to identify any issues in my relatively straightforward implementation.

AlexRadch committed:
Remove unused TestUtilities import in tests

AlexRadch committed:
Tests bugs have been fixed.

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -35,6 +29,16 @@ public static bool Contains(this ReadOnlySpan<char> span, ReadOnlySpan<char> val
+        /// <see langword="false"/>.
+        /// </summary>
+        /// <param name="span">The source span.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Does this need `AggressiveInlining`? It should be consistent with `IsWhiteSpace` since their IL are equivalent in complexity.

huoyaoyuan commented on src/libraries/System.Private.CoreLib/src/System/SearchValues/SearchValues.cs:
```diff
@@ -18,6 +19,27 @@ namespace System.Buffers
+                if (_whiteSpaces is null)
+                {
+                    List<char> whiteSpaces = [];
+                    for (int i = 0; i <= char.MaxValue; i++)
```
This is non-trivial overhead in startup. This should be drive by a table in `CharUnicodeInfo`, which is generated by tool.

MihaZupan replied:
Agreed, since we're in the runtime we can precompute the set and ensure we have sufficient test coverage.
In this case I think it should be
```c#
"\t\n\v\f\r \u0085\u00A0\u1680\u2000\u2001\u2002\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u2028\u2029\u202F\u205F\u3000"
```
and then use a static readonly instead of lazy-init.

MihaZupan replied:
If we're precomputing this, the implementation you're gonna get for `.Contains(char)` is
https://github.com/dotnet/runtime/blob/d2cdcdd69391bbfde1a8fb4aa275a6cc393ca65b/src/libraries/System.Private.CoreLib/src/System/SearchValues/ProbabilisticMapState.cs#L83
We can measure if it makes sense to switch `char.IsWhiteSpace` to using `SearchValues.Contains` here (or even hardcoding the constants if it matters).

MihaZupan commented on src/libraries/System.Memory/tests/Span/StringSearchValues.cs:
```diff
@@ -469,6 +470,254 @@ public static void TestIndexOfAny_RandomInputs()
         }
 
+        [Fact]
+        public static void IsWhiteSpace_True()
```
Please move these tests to https://github.com/dotnet/runtime/tree/main/src/libraries/System.Memory/tests/ReadOnlySpan

`SearchValues` is an implementation detail, and this file is specifically focused on `SearchValues<string>`, which we aren't using here.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -14,15 +15,8 @@ public static partial class MemoryExtensions
-            return true;
-        }
+        public static bool IsWhiteSpace(this ReadOnlySpan<char> span) =>
+            SearchValues.WhiteSpaces.IndexOfAnyExcept(span) < 0;
```
```suggestion
            !SearchValues.WhiteSpaces.ContainsAnyExcept(span);
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -35,6 +29,16 @@ public static bool Contains(this ReadOnlySpan<char> span, ReadOnlySpan<char> val
+        /// <param name="span">The source span.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static bool ContainsAnyWhiteSpace(this ReadOnlySpan<char> span) =>
+            SearchValues.WhiteSpaces.IndexOfAny(span) >= 0;
```
```suggestion
            SearchValues.WhiteSpaces.ContainsAny(span);
```

MihaZupan reviewed (commented):
A concern we had with the implementation here was whether it's worth paying the overhead of taking a vectorized path if we expect the match to commonly be close to the start.

For example it's not obvious why `span.IsWhiteSpace` should use a different strategy compared to `string.IsNullOrWhiteSpace`, where the 99% case is that you're validating an argument and expect it have non-whitespace chars (often already with the first character).

Have you done any perf measurements for such cases?

AlexRadch committed:
Apply suggestions from code review

Co-authored-by: Miha Zupan <mihazupan.zupan1@gmail.com>

AlexRadch committed:
Suggestions from code review

AlexRadch committed:
Remove unused namespaces in SearchValues.cs

AlexRadch commented:
> A concern we had with the implementation here was whether it's worth paying the overhead of taking a vectorized path if we expect the match to commonly be close to the start.
> 
> For example it's not obvious why `span.IsWhiteSpace` should use a different strategy compared to `string.IsNullOrWhiteSpace`, where the 99% case is that you're validating an argument and expect it have non-whitespace chars (often already with the first character).
> 
> Have you done any perf measurements for such cases?

I haven't conducted performance measurements for such cases yet, but I’d be happy to write and run performance tests if we can determine the most relevant data set for the evaluation. Could you clarify what types of inputs or scenarios would best reflect the use cases you're concerned about? For example, should we focus on inputs where the match is typically close to the start, or would you like a broader range of cases? Your guidance will help ensure the tests provide meaningful insights.

AlexRadch commented:
> Have you done any perf measurements for such cases?

I write benchmarks for [the ROSpan IsWhiteSpace method](https://github.com/dotnet/performance/pull/4641).

On my computer, I got the next results
```
| Method       | input                                | Mean      | Error     | StdDev    | Median    | Min       | Max       | Ratio | RatioSD | Allocated | Alloc Ratio |
|------------- |------------------------------------- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|----------:|------------:|
| IsWhiteSpace |                                      |  4.476 ns | 0.1625 ns | 0.1872 ns |  4.497 ns |  4.152 ns |  4.810 ns |  1.00 |    0.06 |         - |          NA |
| IsWhiteSpace |                                      |  5.987 ns | 0.3508 ns | 0.4040 ns |  5.872 ns |  5.272 ns |  6.793 ns |  1.34 |    0.10 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |              (...)    32abcdefg [41] | 54.426 ns | 1.4243 ns | 1.6403 ns | 53.878 ns | 52.365 ns | 58.301 ns |  1.00 |    0.04 |         - |          NA |
| IsWhiteSpace |              (...)    32abcdefg [41] | 18.493 ns | 0.4089 ns | 0.4708 ns | 18.587 ns | 17.435 ns | 19.203 ns |  0.34 |    0.01 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |                 16abcdefg            | 23.507 ns | 0.6364 ns | 0.7329 ns | 23.583 ns | 22.129 ns | 24.579 ns |  1.00 |    0.04 |         - |          NA |
| IsWhiteSpace |                 16abcdefg            | 13.437 ns | 0.4064 ns | 0.4680 ns | 13.304 ns | 12.802 ns | 14.431 ns |  0.57 |    0.03 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |         8abcdefg                     | 14.270 ns | 0.5583 ns | 0.6429 ns | 14.083 ns | 13.210 ns | 16.078 ns |  1.00 |    0.06 |         - |          NA |
| IsWhiteSpace |         8abcdefg                     | 12.534 ns | 0.2626 ns | 0.2457 ns | 12.582 ns | 12.059 ns | 12.842 ns |  0.88 |    0.04 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |        7abcdefg                      | 13.390 ns | 0.6858 ns | 0.7898 ns | 13.449 ns | 11.830 ns | 15.098 ns |  1.00 |    0.08 |         - |          NA |
| IsWhiteSpace |        7abcdefg                      | 13.775 ns | 0.6596 ns | 0.7596 ns | 14.001 ns | 12.358 ns | 14.669 ns |  1.03 |    0.08 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |       6abcdefg                       | 11.988 ns | 0.4449 ns | 0.5123 ns | 12.089 ns | 10.959 ns | 12.698 ns |  1.00 |    0.06 |         - |          NA |
| IsWhiteSpace |       6abcdefg                       | 12.203 ns | 0.2964 ns | 0.3413 ns | 12.160 ns | 11.691 ns | 13.234 ns |  1.02 |    0.05 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |      5abcdefg                        | 10.910 ns | 0.5791 ns | 0.6669 ns | 10.880 ns |  9.743 ns | 12.099 ns |  1.00 |    0.08 |         - |          NA |
| IsWhiteSpace |      5abcdefg                        | 12.345 ns | 0.2731 ns | 0.3145 ns | 12.397 ns | 11.705 ns | 12.870 ns |  1.14 |    0.07 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |     4abcdefg                         |  9.857 ns | 0.2966 ns | 0.3415 ns |  9.885 ns |  9.146 ns | 10.515 ns |  1.00 |    0.05 |         - |          NA |
| IsWhiteSpace |     4abcdefg                         | 12.590 ns | 0.3882 ns | 0.4470 ns | 12.583 ns | 11.711 ns | 13.525 ns |  1.28 |    0.06 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |   2abcdefg                           |  7.499 ns | 0.5524 ns | 0.6362 ns |  7.552 ns |  6.374 ns |  8.743 ns |  1.01 |    0.12 |         - |          NA |
| IsWhiteSpace |   2abcdefg                           | 11.754 ns | 0.5026 ns | 0.5788 ns | 11.648 ns | 10.662 ns | 12.915 ns |  1.58 |    0.15 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace |  1abcdefg                            |  5.908 ns | 0.3721 ns | 0.4285 ns |  5.920 ns |  5.214 ns |  6.715 ns |  1.01 |    0.10 |         - |          NA |
| IsWhiteSpace |  1abcdefg                            | 11.137 ns | 0.4656 ns | 0.5362 ns | 11.094 ns | 10.396 ns | 12.558 ns |  1.89 |    0.16 |         - |          NA |
|              |                                      |           |           |           |           |           |           |       |         |           |             |
| IsWhiteSpace | 0abcdefg                             |  5.080 ns | 0.2560 ns | 0.2949 ns |  5.046 ns |  4.565 ns |  5.709 ns |  1.00 |    0.08 |         - |          NA |
| IsWhiteSpace | 0abcdefg                             | 11.485 ns | 0.2555 ns | 0.2840 ns | 11.450 ns | 11.074 ns | 11.982 ns |  2.27 |    0.14 |         - |          NA |
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/SearchValues/SearchValues.cs:
```diff
@@ -18,6 +18,9 @@ namespace System.Buffers
     /// </remarks>
     public static class SearchValues
     {
+        internal static readonly SearchValues<char> WhiteSpaces = Create(
```
Just remembered that we already have this defined in
https://github.com/dotnet/runtime/blob/5667c2887f607939e5ce054137481f877530a279/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs#L38-L39

AlexRadch committed:
Replaced references to `SearchValues.WhiteSpaces` with
`string.SearchValuesStorage.WhiteSpaceChars` in the
`MemoryExtensions` class methods.

AlexRadch committed:
Remove unused using directive in MemoryExtensions

This commit removes the `using System.Buffers;` directive from the `MemoryExtensions.Globalization.cs` file, indicating that the functionality provided by the `System.Buffers` namespace is no longer needed or has been replaced by alternative implementations.

jeffhandley committed:
Merge branch 'main' into MemoryExtensions.WhiteSpace

PranavSenthilnathan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -35,6 +29,16 @@ public static bool Contains(this ReadOnlySpan<char> span, ReadOnlySpan<char> val
+        /// white-space characters</see>, and returns <see langword="true"/> if found. If not found, returns
+        /// <see langword="false"/>.
+        /// </summary>
+        /// <param name="span">The source span.</param>
```
It's better to match the docs (https://learn.microsoft.com/en-us/dotnet/api/system.memoryextensions.iswhitespace) for the summary/return values:

```suggestion
        /// <summary>
        /// Indicates whether the specified span contains <see cref="char.IsWhiteSpace(char)">any
        /// white-space characters</see>.
        /// </summary>
        /// <param name="span">The source span.</param>
        /// <returns><see langword="true"/> if the span contains any whitespace characters, <see langword="false"/> otherwise.</returns>
```

(Same for the other new methods in this PR.)

huoyaoyuan reviewed (commented):
There are pending comments that need to be resolved.

AlexRadch committed:
Apply suggestion from @PranavSenthilnathan

Co-authored-by: Pranav Senthilnathan <pranav.senthilnathan@live.com>

jeffhandley committed:
Merge branch 'main' into MemoryExtensions.WhiteSpace

PranavSenthilnathan committed:
Merge branch 'main' into MemoryExtensions.WhiteSpace

PranavSenthilnathan committed:
Remove AggressiveInlining and revert IsWhiteSpace

PranavSenthilnathan committed:
Remove extra new line

PranavSenthilnathan committed:
Merge branch 'main' into MemoryExtensions.WhiteSpace

PranavSenthilnathan committed:
Merge branch 'main' into MemoryExtensions.WhiteSpace

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -35,6 +36,15 @@ public static bool Contains(this ReadOnlySpan<char> span, ReadOnlySpan<char> val
+        /// white-space characters</see>.
+        /// </summary>
+        /// <param name="span">The source span.</param>
+        /// <returns><see langword="true"/> if the span contains any whitespace characters, <see langword="false"/> otherwise.</returns>
```
```suggestion
        /// Indicates whether the specified span contains any <see cref="char.IsWhiteSpace(char)">
        /// white-space characters</see>.
        /// </summary>
        /// <param name="span">The source span.</param>
        /// <returns><see langword="true"/> if the span contains any white-space characters, <see langword="false"/> otherwise.</returns>
```
Nit: we should try to at least be locally consistent with white-space vs whitespace.

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -149,6 +159,22 @@ public static int IndexOf(this ReadOnlySpan<char> span, ReadOnlySpan<char> value
 
+        /// <summary>
+        /// Reports the zero-based index of the first occurrence of <see cref="char.IsWhiteSpace(char)">any white-space
+        /// characters</see> in the current <paramref name="span"/>, or -1 if not founded.
```
```suggestion
        /// Reports the zero-based index of the first occurrence of any <see cref="char.IsWhiteSpace(char)">
        /// white-space character</see> in the current <paramref name="span"/>, or -1 if not found.
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -149,6 +159,22 @@ public static int IndexOf(this ReadOnlySpan<char> span, ReadOnlySpan<char> value
+
+        /// <summary>
+        /// Reports the zero-based index of the first occurrence of <see cref="char.IsWhiteSpace(char)">any
+        /// non-white-space characters</see> in the current <paramref name="span"/>, or -1 if not founded.
```
```suggestion
        /// Reports the zero-based index of the first occurrence of any <see cref="char.IsWhiteSpace(char)">
        /// non-white-space character</see> in the current <paramref name="span"/>, or -1 if not found.
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -184,6 +210,22 @@ ref MemoryMarshal.GetReference(value),
 
+        /// <summary>
+        /// Reports the zero-based index of the last occurrence of <see cref="char.IsWhiteSpace(char)">any white-space
+        /// characters</see> in the current <paramref name="span"/>, or -1 if not founded.
```
```suggestion
        /// Reports the zero-based index of the last occurrence of any <see cref="char.IsWhiteSpace(char)">
        /// white-space character</see> in the current <paramref name="span"/>, or -1 if not found.
```

MihaZupan commented on src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Globalization.cs:
```diff
@@ -184,6 +210,22 @@ ref MemoryMarshal.GetReference(value),
+
+        /// <summary>
+        /// Reports the zero-based index of the last occurrence of <see cref="char.IsWhiteSpace(char)">any
+        /// non-white-space characters</see> in the current <paramref name="span"/>, or -1 if not founded.
```
```suggestion
        /// Reports the zero-based index of the last occurrence of any <see cref="char.IsWhiteSpace(char)">
        /// non-white-space character</see> in the current <paramref name="span"/>, or -1 if not found.
```

MihaZupan committed:
Improve comments

Co-authored-by: Miha Zupan <mihazupan.zupan1@gmail.com>
